src/company_gui.cpp

```diff
@@ -1827,267 +1845,217 @@ struct CompanyInfrastructureWindow : Window
+
+	void SetStringParameters(WidgetID widget) const override
+	{
+		switch (widget) {
+			case WID_CI_CAPTION:
+				SetDParam(0, (CompanyID)this->window_number);
 				break;
-			}
 		}
 	}
 
-	/**
-	 * Helper for drawing the counts line.
-	 * @param r            The bounds to draw in.
-	 * @param y            The y position to draw at.
-	 * @param count        The count to show on this line.
-	 * @param monthly_cost The monthly costs.
-	 */
-	void DrawCountLine(const Rect &r, int &y, int count, Money monthly_cost) const
+	inline int GetLineHeight() const
+	{
+		return GetCharacterHeight(FS_NORMAL) + WidgetDimensions::unscaled.vsep_normal / 2;
+	}
+
+	void UpdateWidgetSize(WidgetID widget, Dimension &size, [[maybe_unused]] const Dimension &padding, [[maybe_unused]] Dimension &fill, [[maybe_unused]] Dimension &resize) override
 	{
-		SetDParam(0, count);
-		DrawString(r.left, r.right, y += GetCharacterHeight(FS_NORMAL), STR_JUST_COMMA, TC_WHITE, SA_RIGHT);
+		if (widget != WID_CI_LIST) return;
+
+		uint  max_lwidth = 0;
+		uint  max_count  = 10000;    // Some random number to reserve enough space.
+		Money max_cost   = 1000000;   // Some random number to reserve enough space.
+
+		for (const auto &ie : this->infrastructurelist) {
+			max_lwidth = std::max(max_lwidth, GetStringBoundingBox(ie.label).width);
+			max_cost   = std::max(max_cost, ie.monthly_cost);
+			max_count  = std::max(max_count, ie.count);
+		}
+
+		/* add ident to labels */
+		max_label_width = max_lwidth + WidgetDimensions::scaled.hsep_indent;
+
+		SetDParam(0, max_count);
+
+		max_count_width = GetStringBoundingBox(STR_JUST_COMMA).width + padding.width * 2;
+		size.height     = static_cast<uint>(this->infrastructurelist.size() * GetLineHeight());
 
 		if (_settings_game.economy.infrastructure_maintenance) {
-			SetDParam(0, monthly_cost * 12); // Convert to per year
-			Rect tr = r.WithWidth(this->total_width, _current_text_dir == TD_RTL);
-			DrawString(tr.left, tr.right, y,
-				TimerGameEconomy::UsingWallclockUnits() ? STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_PERIOD : STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_YEAR,
-				TC_FROMSTRING, SA_RIGHT);
+			SetDParam(0, max_cost * 12);
+			max_cost_width = GetStringBoundingBox(TimerGameEconomy::UsingWallclockUnits()
+					? STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_PERIOD : STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_YEAR).width;
+
+			size.height   += WidgetDimensions::unscaled.vsep_normal * 2;
+		} else {
+			max_cost_width = 0;
 		}
+
+		size.height   = Clamp(size.height, 0, GetLineHeight() * 25);   // max 25 lines otherwise expand widow manually
+		size.width    = max_label_width + WidgetDimensions::scaled.hsep_normal + max_cost_width + WidgetDimensions::scaled.hsep_normal + max_count_width;
+
+		resize.height = GetLineHeight();
 	}
 
 	void DrawWidget(const Rect &r, WidgetID widget) const override
 	{
-		const Company *c = Company::Get((CompanyID)this->window_number);
+		if (widget != WID_CI_LIST) return;
 
-		int y = r.top;
+		int y = r.top + 1;
```

Comment:
+ 1 here seems wrong. If padding is needed then it should be one of the standard scaled WidgetDimensions.